### PR TITLE
Acceptance test to view credential

### PR DESCRIPTION
**Pull request by awais786**
For theses tests data should be exists in lms and programs. 

1. In LMS user should have completed x-series program.

2. In test case it will try to try to credential. If test case already exist then will change in db. Then test in a test user will login on lms and from dashboard it will click the credential link.  

3. In other test case first credentials will be revoked and then again user will login and in this case there will be no credential.

### CODE REVIEW

**Comment by awais786** [comment]
@jimabramson Please review this PR.

**Comment by ahsan-ul-haq** [comment]
:+1:

**Comment by awais786** [comment]
@rlucioni Please review this PR.

File: acceptance_tests/mixins.py [hunk 1]
```diff
@@ -21,3 +25,44 @@ def login_with_lms(self):
 
         self.lms_login_page.browser.get(self.lms_login_page.url())  # pylint: disable=not-callable
         self.lms_login_page.login(email, password)
+
+
+class CredentialsApiMixin(object):
+    """ Mixin used for login on credentials."""
+    def setUp(self):
+        super(CredentialsApiMixin, self).setUp()
+        self.data = ''
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: I'd initialize this to `None` instead of the empty string.

File: acceptance_tests/mixins.py [hunk 2]
```diff
@@ -21,3 +25,44 @@ def login_with_lms(self):
...
+    def _create_credential(self):
+        """Create user credential for a program."""
+
+        return self.credential_api_client.user_credentials.post({
+            'username': LMS_USERNAME,
+            'credential': {'program_id': PROGRAM_ID},
+            'attributes': []
+        })
+
+    def _change_credential_status(self, data, status):
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: Use `self.data` instead of taking it as an argument.

File: acceptance_tests/mixins.py [hunk 3]
```diff
@@ -21,3 +25,44 @@ def login_with_lms(self):
...
+
+
+class CredentialsApiMixin(object):
+    """ Mixin used for login on credentials."""
+    def setUp(self):
+        super(CredentialsApiMixin, self).setUp()
+        self.data = ''
+
+    @property
+    def credential_api_client(self):
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: I'd like to see this client use an OAuth 2.0 access token instead of a JWT. It'll be cleaner. For one, you won't need a signing key nor an OAuth client ID to run the tests.

In order to do this, you'll need to implement and use an authentication class similar to [BearerAuthentication](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/api/authentication.py#L81) from ecommerce.

File: acceptance_tests/mixins.py [hunk 4]
```diff
@@ -21,3 +25,44 @@ def login_with_lms(self):
...
+        payload = {
+            "iss": OAUTH_URL,
+            "aud": 'credentials-key',
+            "exp": now + datetime.timedelta(seconds=expires_in),
+            "iat": now,
+            "preferred_username": LMS_USERNAME,
+            "administrator": True,
+        }
+        try:
+            jwt_data = jwt.encode(payload, 'credentials-secret')
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: Signing keys and other configuration should not be hardcoded here. This should go away when you switch to using an access token.

File: acceptance_tests/test_rendering.py [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+from acceptance_tests.mixins import LoginMixin, CredentialsApiMixin
+from acceptance_tests.pages import LMSDashboardPage
+
+
+@skipUnless(ENABLE_OAUTH2_TESTS, 'OAuth2 tests are not enabled.')
+class RenderCredentialTests(LoginMixin, WebAppTest, CredentialsApiMixin):
+
+    def setUp(self):
+        super(RenderCredentialTests, self).setUp()
+        self.browser.maximize_window()
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: Why is this necessary?

File: acceptance_tests/test_rendering.py [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+        """
+        self.login_with_lms()
+        lms = LMSDashboardPage(self.browser).wait_for_page()
+        lms.is_browser_on_page()
+
+        self.assertTrue(lms.credential_available())
+
+        credential_link = lms.get_credential_link()
+
+        # click the credential link
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: This comment is unnecessary.

File: acceptance_tests/test_rendering.py [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+from selenium.webdriver.common.by import By
+from selenium.webdriver.support import expected_conditions as EC
+from selenium.webdriver.support.ui import WebDriverWait
+
+from acceptance_tests.config import ENABLE_OAUTH2_TESTS
+from acceptance_tests.mixins import LoginMixin, CredentialsApiMixin
+from acceptance_tests.pages import LMSDashboardPage
+
+
+@skipUnless(ENABLE_OAUTH2_TESTS, 'OAuth2 tests are not enabled.')
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: Don't do this. The acceptance tests which already exist in this repo are decorated like this, but they shouldn't be. It doesn't make sense if you look at what's going on in the config module. If an access token isn't provided, the tests won't run.

File: acceptance_tests/test_rendering.py [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+        self.data = self._create_credential()
+        self._change_credential_status(self.data, 'awarded')
+
+    def test_student_dashboard_with_certificate(self):
+        """ Render the credentials link from student dashboard.
+        After successful login user will redirect towards the dashboard page. and
+        click on the credential link.
+        """
+        self.login_with_lms()
+        lms = LMSDashboardPage(self.browser).wait_for_page()
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: This would be much easier to read if the variable was named `dashboard` or `student_dashboard`.

File: acceptance_tests/mixins.py (hunk 3 above)
Comment (awais786) [inline acceptance_tests/mixins.py]: @rlucioni can we implement this into another story ?. 
Because It will increase the scope. I have created another PR for this https://github.com/edx/credentials/pull/67

Comment (rlucioni) [inline acceptance_tests/mixins.py]: Implementing this in a separate PR is great. Are you planning to finish up #67, merge it, update these tests, then merge this PR?

Comment (awais786) [inline acceptance_tests/mixins.py]: I have a plan to merge tests and then merge other PR and then fix tests again. What do u think ?

Comment (awais786) [inline acceptance_tests/mixins.py]: @rlucioni I think first merging these test will be a good plan.

**Comment by awais786** [comment]
@rlucioni Please review this PR again. BearerAuth will be implemented in other story. We can merged this PR without BearerAuth.

File: acceptance_tests/mixins.py [hunk 9]
```diff
@@ -21,3 +27,44 @@ def login_with_lms(self):
...
+        }
+        try:
+            jwt_data = jwt.encode(payload, JWT_SECRET_KEY)
+            api_client = EdxRestApiClient(CREDENTIALS_API_URL, jwt=jwt_data)
+        except Exception:  # pylint: disable=broad-except
+            log.exception("Failed to initialize the API client with url '%s'.", CREDENTIALS_API_URL)
+            return
+        return api_client
+
+    def _create_credential(self):
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: This method and the one below aren't really private, are they? I would drop the underscores.

File: acceptance_tests/mixins.py [hunk 10]
```diff
@@ -21,3 +27,44 @@ def login_with_lms(self):
...
+        """Create user credential for a program."""
+
+        return self.credential_api_client.user_credentials.post({
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: Instead of returning data only to update the `data` class attribute later, just update `self.data` here. It'll make your code a little cleaner.

File: acceptance_tests/test_rendering.py [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+    def setUp(self):
+        super(RenderCredentialTests, self).setUp()
+        self.data = self._create_credential()
+        self._change_credential_status('awarded')
+
+    def test_student_dashboard_with_certificate(self):
+        """ Render the credentials link from student dashboard.
+        After successful login user will redirect towards the dashboard page. and
+        click on the credential link.
+        """
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: I'd clean this up a bit:
```python
"""Verify that links to XSeries credentials are displayed on the student dashboard."""
```

File: acceptance_tests/pages.py [hunk 12]
```diff
@@ -54,9 +54,22 @@ def login(self, username, password):
...
+        self.credential_css_selector = '.wrapper-xseries-certificates'
+
     @property
     def url(self):
         return self._build_url('dashboard')
 
     def is_browser_on_page(self):
         return self.browser.title.startswith('Dashboard')
+
+    def credential_available(self):
```
Comment (rlucioni) [inline acceptance_tests/pages.py]: Give this method a name more consistent with what it does. For example, `are_credential_links_present`.

File: acceptance_tests/pages.py [hunk 13]
```diff
@@ -54,9 +54,22 @@ def login(self, username, password):
...
+        return self.q(css=self.credential_css_selector).present
+
+    def credential_link(self):
```
Comment (rlucioni) [inline acceptance_tests/pages.py]: Same as above. I'd use `click_credential_link`.

File: acceptance_tests/test_rendering.py [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+        student_dashboard.credential_link()
+        WebDriverWait(self.browser, 10).until(EC.presence_of_element_located((By.ID, 'action-print-view')))
+
+        # check the credential hash-id matches with rendered credential page.
+        self.assertIn(
+            self.browser.find_elements_by_css_selector('.accomplishment-stamp-validity span:nth-of-type(2)')[0].text,
+            credential_link
+        )
+
+
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: Remove this line. Within classes, only one line should separate methods.

File: acceptance_tests/test_rendering.py [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+    def test_student_dashboard_without_certificate(self):
+        """ If student has no credentials or with revoked status then no link
```
Comment (rlucioni) [inline acceptance_tests/test_rendering.py]: This test doesn't cover the case where the student has no credentials, only the case in which credentials are revoked. Either update the docstring or update the test so that the credential is deleted and the dashboard is re-checked.

**Comment by awais786** [comment]
@rlucioni it's ready for final review.

File: acceptance_tests/mixins.py [hunk 16]
```diff
@@ -1,6 +1,13 @@
...
 import logging
 
-from acceptance_tests.config import LMS_EMAIL, LMS_PASSWORD
+from edx_rest_api_client.client import EdxRestApiClient
+import jwt
+
+from acceptance_tests.config import (
+    LMS_EMAIL, LMS_PASSWORD, CREDENTIALS_API_URL, LMS_USERNAME, PROGRAM_ID, OAUTH_URL,
+    JWT_SECRET_KEY, USER_JWT_AUDIENCE
+)
```
Comment (rlucioni) [inline acceptance_tests/mixins.py]: Instead of importing all these constants, just import `config` and access these constants as you need them. It makes it clearer where these constants are coming from.

**Comment by rlucioni** [comment]
:+1: once final comment addressed and commits squashed.

**Comment by awais786** [comment]
Thanks @rlucioni I will do this change tomorrow and than merge it.
